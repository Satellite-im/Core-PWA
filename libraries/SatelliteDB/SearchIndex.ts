import MiniSearch, { Options } from 'minisearch'
import type { SearchOptions } from 'minisearch'

export type SearchIndexOptions = {
  schema: Options<any>
  records?: any[]
  index?: MiniSearch
}

export default class SearchIndex {
  index: MiniSearch
  schema: Options<any>
  documents: Map<string, any> = new Map()

  constructor(
    {
      records = [],
      schema = { fields: ['id'] },
      index = undefined,
    }: SearchIndexOptions = { schema: { fields: ['id'] } },
  ) {
    this.schema = schema
    if (index) {
      this.index = index
    } else {
      this.index = new MiniSearch(this.schema)
    }
    if (records) this.index.addAll(records)
  }

  get idField() {
    return this.schema.idField || 'id'
  }

  add(record: any) {
    this.documents.set(record[this.idField], record)
    this.index.add(record)
  }

  addAll(records: any[]) {
    records.map((record) => this.add(record))
  }

  getById(id: string) {
    return this.documents.get(id)
  }

  upsert(record: any) {
    this.remove(record)
    this.add(record)
  }

  upsertAll(records: any[]) {
    return records.map((record) => this.upsert(record))
  }

  removeAll(records?: any[]) {
    return (records || Array.from(this.documents.values())).map((record) =>
      this.remove(record),
    )
  }

  remove(record: any) {
    if (this.documents.has(record[this.idField])) {
      const doc = this.documents.get(record[this.idField])
      this.index.remove(doc)
      this.documents.delete(record[this.idField])
    }
  }

  update(records: any[]) {
    this.documents = new Map()
    this.index.removeAll()
    this.addAll(records)
  }

  search(query: string, options?: SearchOptions | undefined) {
    return this.index
      .search(query, options)
      .map((doc) => ({ ...doc, ...this.getById(doc[this.idField]) }))
  }

  autoSuggest(query: string, options?: SearchOptions | undefined) {
    return this.index.autoSuggest(query, options)
  }

  serialize(): string {
    const serialized = JSON.stringify({
      index: this.index.toJSON(),
      schema: this.schema,
    })
    return serialized
  }

  static deserialize(serialized: string) {
    const { index, schema } = JSON.parse(serialized)
    const idx = MiniSearch.loadJS(index, schema)
    return new SearchIndex({ index: idx, schema })
  }
}
