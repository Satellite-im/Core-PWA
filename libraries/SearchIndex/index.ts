import lunr from 'lunr'

export type SearchIndexSchema = {
  ref: string
  fields: string[]
}

/**
 * Class representing a SearchIndex
 * @class SearchIndex
 */
export default class SearchIndex {
  schema: SearchIndexSchema
  index?: lunr.Index

  /**
   * @constructs SearchIndex
   * @param schema the schema to use for the index
   */
  constructor(schema: SearchIndexSchema) {
    this.schema = schema
  }

  /**
   * Builds a search index from a list of documents.
   * @param schema the schema of the index
   * @param data the data to index
   * @returns lunr.Index
   */
  static build(schema: SearchIndexSchema, data: any[]) {
    return lunr(function () {
      this.ref(schema.ref)
      schema.fields.forEach((field) => this.field(field))
      data.forEach((record) => this.add(record))
    })
  }

  /**
   * Updates the index with new data.
   * @param data the data to index
   */
  update(data: any[]) {
    this.index = SearchIndex.build(this.schema, data)
  }

  /**
   * Serializes the index to JSON.
   * @returns JSON
   */
  serialize() {
    return JSON.stringify({
      schema: this.schema,
      index: this.index,
    })
  }

  /**
   * Deserialize an index from JSON.
   * @param serialized the serialized index
   * @returns SearchIndex
   */
  static deserialize(serialized: string) {
    const { schema, index } = JSON.parse(serialized)
    const searchIndex = new SearchIndex(schema)
    searchIndex.index = index
    return searchIndex
  }

  /**
   * Searches the index for a query string.
   * @param query the query to search for
   * @param retry is this a retry?
   * @returns an array of results
   */
  search(query: string, retry = false): lunr.Index.Result[] | undefined {
    try {
      return this.index?.search(query)
    } catch (err) {
      if (retry) {
        console.warn('invalid search query', err)
        return []
      }
      // retry without punctuation
      return this.search(query.replace(/[^\w\s]/gi, ''), true)
    }
  }

  /**
   * Subscribe to an observable that emits data to be indexed
   * @param observable the observable to subscribe to
   * @param callback the callback to call when data is emitted
   */
  subscribe(observable: any, callback: any) {
    return observable.subscribe((data: any[]) => {
      this.update(data)
      if (callback) callback(data)
    })
  }

  /**
   * Unsubscribe from an observable that emits data to be indexed
   * @param observable the observable to unsubscribe from
   */
  unsubscribe(observable: any) {
    return observable.unsubscribe()
  }
}
