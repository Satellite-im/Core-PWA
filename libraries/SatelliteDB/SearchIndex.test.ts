import SearchIndex from './SearchIndex'

describe('SatelliteDB/SearchIndex', () => {
  let idx: SearchIndex
  beforeAll(() => {
    idx = new SearchIndex({
      schema: {
        fields: ['id', 'text'],
        storeFields: ['id', 'text'],
      },
    })
  })

  beforeEach(() => {
    idx.update([])
  })

  test('constructor without a schema', () => {
    const idxb = new SearchIndex()
  })

  test('searchIndex.update()', async () => {
    const data = [
      { id: '1', text: 'first match' },
      { id: '2', text: 'second match' },
      { id: '3', text: 'third match' },
    ]
    idx.update(data)
    expect(idx.search('first')?.map((r) => r.id)).toEqual(['1'])
    expect(idx.search('second')?.map((r) => r.id)).toEqual(['2'])
    expect(idx.search('third')?.map((r) => r.id)).toEqual(['3'])
    expect(idx.search('match')?.map((r) => r.id)).toEqual(['3', '2', '1'])
  })

  test('searchIndex.update()', async () => {
    const data = [
      { id: '1', text: 'first match' },
      { id: '2', text: 'second match' },
      { id: '3', text: 'third match' },
    ]
    idx.update(data)
    expect(idx.autoSuggest('first')).toMatchSnapshot()
  })

  test('searchIndex.update() without an id', async () => {
    const data = [{ text: 'foo bar' }]
    expect(() => idx.update(data)).toThrow()
  })

  test('searchIndex.add()', async () => {
    const data = { id: '1', text: 'foo bar' }
    idx.add(data)
    expect(idx.search('foo')?.map((r) => r.id)).toEqual(['1'])
  })

  test('searchIndex.add() without an id', async () => {
    const data = { text: 'foo bar' }
    expect(() => idx.add(data)).toThrow()
  })

  test('searchIndex.remove()', async () => {
    const data = [
      { id: '1', text: 'first match' },
      { id: '2', text: 'second match' },
    ]
    idx.update(data)
    expect(idx.search('first')?.map((r) => r.id)).toEqual(['1'])
    expect(idx.search('second')?.map((r) => r.id)).toEqual(['2'])
    expect(idx.search('match')?.map((r) => r.id)).toEqual(['2', '1'])
    idx.remove(data[0])
    expect(idx.search('first')?.map((r) => r.id)).toEqual([])
    expect(idx.search('match')?.map((r) => r.id)).toEqual(['2'])
  })

  test('searchIndex.addAll()', async () => {
    const data = [
      { id: '1', text: 'first match' },
      { id: '2', text: 'second match' },
    ]
    idx.addAll(data)
    expect(idx.search('first')?.map((r) => r.id)).toEqual(['1'])
    expect(idx.search('second')?.map((r) => r.id)).toEqual(['2'])
    expect(idx.search('match')?.map((r) => r.id)).toEqual(['2', '1'])
  })

  test('searchIndex.removeAll()', async () => {
    const data = [
      { id: '1', text: 'first match' },
      { id: '2', text: 'second match' },
    ]
    idx.update(data)
    idx.removeAll()
    expect(idx.search('first')?.map((r) => r.id)).toEqual([])
    expect(idx.search('second')?.map((r) => r.id)).toEqual([])
    expect(idx.search('match')?.map((r) => r.id)).toEqual([])
  })

  test('searchIndex.upsert()', async () => {
    const data = { id: '1', text: 'foo bar' }
    idx.upsert(data)
    idx.upsert({ id: '1', text: 'foo bar baz' })
    expect(idx.search('foo')?.length).toEqual(1)
    expect(idx.search('foo')?.[0]?.text).toEqual('foo bar baz')
  })

  test('searchIndex.upsertAll()', async () => {
    const data = [
      { id: '1', text: 'foo bar' },
      { id: '1', text: 'foo bar baz' },
    ]
    idx.upsertAll(data)
    expect(idx.search('foo')?.map((r) => r.text)).toEqual(['foo bar baz'])
  })
})
