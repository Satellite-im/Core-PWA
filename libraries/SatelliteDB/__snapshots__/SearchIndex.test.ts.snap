// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SatelliteDB/SearchIndex searchIndex.update() 1`] = `
Array [
  Object {
    "score": 1.0986122886681096,
    "suggestion": "first",
    "terms": Array [
      "first",
    ],
  },
]
`;
